pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/QueryDispatcher.java

```diff
@@ -146,7 +146,13 @@ public QueryResult submitAndReduce(RequestContext context, DispatchableSubPlan d
     Set<QueryServerInstance> servers = new HashSet<>();
     try {
       submit(requestId, dispatchableSubPlan, timeoutMs, servers, queryOptions);
-      return runReducer(requestId, dispatchableSubPlan, timeoutMs, queryOptions, _mailboxService);
+      try {
+        return runReducer(requestId, dispatchableSubPlan, timeoutMs, queryOptions, _mailboxService);
+      } finally {
+        if (isQueryCancellationEnabled()) {
+          _serversByQuery.remove(requestId);
+        }
+      }
```

Comment:
IMO Given it is submit the method that adds the query into the map, we should remove the state there as well.

Reply:
I'm not sure I follow. Won't that be too soon to remove it? Submit is only submitting the stage plans to the servers. The reducer is where we're waiting for the result blocks from the final stage. Queries should be cancellable until they finish executing (either successfully or unsuccessfully), so we can't really remove state from `_serversByQuery` in `submit` right?

Reply:
Another option is to move up the state addition into this method as well (i.e., populating `_serversByQuery`).

Reply:
@yashmayya : to Gonzalo's point, can we just use finally with the outer try? If the submit call throws for some reason after the servers were added to the map, that'd still be a leak?

Reply:
@ankitsultana the outer try already has a catch block that ensures that the map entry for the query is removed - https://github.com/apache/pinot/blob/07785bf8cff073791492f2cb1667e2d12709703b/pinot-query-runtime/src/main/java/org/apache/pinot/query/service/dispatch/QueryDispatcher.java#L150-L154

In fact, all exceptional situations seem to already be covered. It's only the successful case that was causing the leak.